functable.h

```diff
@@ -6,12 +6,285 @@
 #ifndef FUNCTABLE_H_
 #define FUNCTABLE_H_
 
+#include "zbuild.h"
+#include "zendian.h"
 #include "deflate.h"
 #include "crc32_fold_c.h"
 #include "adler32_fold_c.h"
 #include "inftrees.h"
 #include "inflate.h"
 
+
+#if defined(UNALIGNED_OK) && BYTE_ORDER == LITTLE_ENDIAN
```

Comment:
I think we should see what all these changes look like if they were moved into their associated arch/x86/*.h headers. Perhaps you can add it as a separate commit, that way it can be easily backed out or squashed at the end. But I think it might clean all this up a bit more.

Reply:
I think it is a bad idea.
Currently there are two dispatching points:
``functable.h`` - for compile-time
``functable.c`` - for runtime.

Moving ``#define`` into the ``arch/x86/*.h`` headers will require multiple checks of the same conditions in headers.

Reply:
Well, I mean for non-x86 those functions would be moved into the header for their own arch folder too.  All the arch ifdef logic would be kept in x86_features or arm_features.h ...

Reply:
@nmoinvaz 
I understood your idea!

Reply:
lol, I had just written a big comment about there probably being a misunderstanding between you two, and trying to explain it better. Was about to press send when your new comment popped up 😄 

You still seem to think it is a problem though, so please explain what the problem is, because I don't see it currently.

---

I don't think that would increase the compiler load by a noticeable degree, so I am wondering whether @phprus misunderstood the suggestion..?

Basically @nmoinvaz suggested splitting this file into arch-dependent files, only one of those would ever be active for a compiler to see.
The C-fallback parts could also move to arch/generic/ where those functions currently live.

Then this file would mainly be the struct and a few other bits and pieces that are functable-specific.

In that case this file would include the generic header file and the arch-specific header file.
If that is still a problem, I am afraid I do not understand what the problem you are describing is @phprus?

Reply:
Splitted.

Reply:
Sorry to be a pain. Anyways, it looks good. Now future changes to x86 code should even more so only touch arch/x86 files, same for other archs too.

Reply:
We need split the runtime dispatcher too, but this is more difficult task.

I didn't succeed, because the ``*_features.h`` headers contain a lot of non-obvious dependencies.
For example:

```c
#ifdef CPU_FEATURES_H_
```

or 
```c
#ifdef INFLATE_H_
```

Reply:
Another implementation of this PR: #1659

Main differences:
- Separate CPU-specific function prototypes and CPU-arch features checker.
- Reducing dependencies between header files.

Separating headers reduces dependencies on implementations (check internal macro ``CPU_FEATURES_H_`` or ``INFLATE_H_ ``, ...)